### DISCUSSION TITLE
[Support]: Performance issues in Firefox

### DISCUSSION THREAD

**Discussion by h3x1e** [blakeblackshear/frigate#11287] [General Support]
### Describe the problem you are having

With the recent update (from what I can remember) when viewing streams either within the Frigate Web UI (mse) or within the Frigate-Card for Home Assistant (go2rtc live view provider) I seem to be getting performance issues.

When live viewing a camera my GPU (or CPU if Hawardware Aceleration is off) ramps up to 3D clock speeds (1905MHz) drawing 113 watts of power and causes Firefox to become slugish, making other video (Youtube / Twitch) to drop frames until I stop viewing the camera.

The issue does not seem to happen on Edge Browser, using Hardware Acceleraction GPU doesnt ramp stays at 2D clocks and 46 watts power draw.

### Version

0.13.2-6476f8a

### Frigate config file

```yaml
mqtt:
  enabled: true
  host: 192.168.0.27
  port: 1883
  user: {username}
  password: {password}
  topic_prefix: frigate

ffmpeg:
  hwaccel_args: preset-vaapi

detectors:
  coral:
    type: edgetpu
    device: usb

record:
  enabled: true
  events:
    retain:
      default: 30
      mode: motion

snapshots:
  enabled: true
  timestamp: true
  retain:
    default: 30

detect:
  enabled: true
  width: 1280
  height: 720
  fps: 5
  stationary:
    interval: 100
    threshold: 50

motion:
  threshold: 40
  
timestamp_style:
  position: bl

go2rtc:
  streams:
    front:
      - rtsp://{USERNAME}:{FRIGATE_RTSP_PASSWORD}@192.168.0.30/stream1      
    front_low:
      - rtsp://{USERNAME}:{FRIGATE_RTSP_PASSWORD}@192.168.0.30/stream2          
    cat_cam:
      - rtsp://{USERNAME}:{FRIGATE_RTSP_PASSWORD}@192.168.0.31/stream1      
    cat_cam_low:
      - rtsp://{USERNAME}:{FRIGATE_RTSP_PASSWORD}@192.168.0.31/stream2

cameras:
  front:
    ffmpeg:        
      inputs:
      - path: rtsp://127.0.0.1:8554/front
        input_args: preset-rtsp-restream
        roles:
          - record
      - path: rtsp://127.0.0.1:8554/front_low
        input_args: preset-rtsp-restream
        roles:
          - detect      

  cat_cam:
    ffmpeg:
      inputs:
      - path: rtsp://127.0.0.1:8554/cat_cam
        input_args: preset-rtsp-restream
        roles:
          - record
      - path: rtsp://127.0.0.1:8554/cat_cam_low
        input_args: preset-rtsp-restream
        roles:
          - detect
    objects:
      track:
        - person
        - cat
        - dog
      filters:
        person:
          mask:
            - 1194,359,1173,541,1114,720,957,712,989,585,1001,513,1027,463,1064,388,1114,352
```

### Relevant log output

```shell
2024-05-08 10:52:08.959308349  [INFO] Preparing Frigate...
2024-05-08 10:52:08.982422516  [INFO] Starting Frigate...
2024-05-08 10:52:10.063390633  [2024-05-08 10:52:10] frigate.app                    INFO    : Starting Frigate (0.13.2-6476f8a)
2024-05-08 10:52:10.097410688  [2024-05-08 10:52:10] peewee_migrate.logs            INFO    : Starting migrations
2024-05-08 10:52:10.101408775  [2024-05-08 10:52:10] peewee_migrate.logs            INFO    : There is nothing to migrate
2024-05-08 10:52:10.106332376  [2024-05-08 10:52:10] frigate.app                    INFO    : Recording process started: 296
2024-05-08 10:52:10.108655869  [2024-05-08 10:52:10] frigate.app                    INFO    : go2rtc process pid: 89
2024-05-08 10:52:10.132879482  [2024-05-08 10:52:10] detector.coral                 INFO    : Starting detection process: 306
2024-05-08 10:52:12.780128786  [2024-05-08 10:52:10] frigate.app                    INFO    : Output process started: 308
2024-05-08 10:52:12.780265827  [2024-05-08 10:52:10] frigate.app                    INFO    : Camera processor started for front: 316
2024-05-08 10:52:12.780497740  [2024-05-08 10:52:10] frigate.app                    INFO    : Camera processor started for cat_cam: 317
2024-05-08 10:52:12.780662277  [2024-05-08 10:52:10] frigate.app                    INFO    : Capture process started for front: 319
2024-05-08 10:52:12.780869190  [2024-05-08 10:52:10] frigate.app                    INFO    : Capture process started for cat_cam: 322
2024-05-08 10:52:12.787406760  [2024-05-08 10:52:10] frigate.detectors.plugins.edgetpu_tfl INFO    : Attempting to load TPU as usb
2024-05-08 10:52:12.787953805  [2024-05-08 10:52:12] frigate.detectors.plugins.edgetpu_tfl INFO    : TPU found
2024-05-08 10:52:57.135079258  [2024-05-08 10:52:57] frigate.comms.dispatcher       INFO    : Turning off detection for cat_cam
```

### FFprobe output from your camera

```shell
Ffprobe Output
Stream 0:
Return Code: 0

Video:

Codec: H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
Resolution: 2560x1440
FPS: 15/1

Stream 1:
Return Code: 0

Video:
```

### Operating system

Debian

### Install method

Docker Compose

### Coral version

USB

### Network connection

Wired

### Camera make and model

Tapo C320WS

### Any other information that may be helpful

Everything seems fine with my config, everything loads and plays as eexpected apart from this issue in Firefox when live viewing any camera. I have also tried a fresh install of Firefox in case it was an addon causing the issue but nothing changed.

**Comment by NickM-27** [comment]
are you talking about the GPU on the client or the host? this sounds like a firefox issue decoding the streams

**Comment by h3x1e** [comment]
Firefox on the Client, yeah. It's weird because I cant remember / didnt notice it happening before the recent Frigate update, but I might be wrong, and there hasnt been a firefox update for a bit iirc.

**Comment by noloboy** [comment]
Did you ever figure this out?  Whenever I use Firefox or any firefox based browsers like Waterfox while viewing the main page or homepage of Frigate (I use v.14) - my Desktop PC uses 35% of CPU usage just for the browser.  Viewing the exact same thing in Chrome is like 2%.   

So my PC tends to spin up all the fans when viewing Frigate homepage.

I have Unifi Protect which also has the exact same cameras, and viewing all cameras on the Protect web page is fine.

**Comment by h3x1e** [comment]
na it's still happening, as Nick said it's probably a Firefox issue. 
I've just been using go2rtc in Home Assistant with Frigate Cards, as a workaround.

**Comment by tenpaiyomi** [comment]
Adding to this as well, I'm experiencing the same behavior.  Load up Frigate in Firefox, GPU video decode jumps up to 90-100%, almost constant, generally regardless of how many cameras are currently displaying a live view, unless the number is 0.  Load up Frigate in a Chrome based browser (Brave, in my case) and I barely crack 10% on average, 20-30% if 6 or 7 of my cameras are showing a live feed at the moment.

I am hesitant to prescribe this to _explicitly_ a Firefox bug, because this only seems to happen with the Frigate page, and only in Firefox.  The streams in Home Assistant don't show any issue, if I have a video playing in YouTube or Twitch or Instagram or wherever, they don't show an issue.  Thus far the only place where I've experienced this sort of performance hit is Frigate.  As to what actually causes this variance, unfortunately I'm not seeing anything that jumps out at me.

**Reply by hawkeye217** [reply]
I use Firefox as my daily driver and I develop many Frigate features with Firefox as the primary browser, and I don't see any performance issues. 

If I had to guess, the issue is probably a combination of your specific hardware and the way Firefox does stream decoding. Frigate uses standard [MSE API calls](https://developer.mozilla.org/en-US/docs/Web/API/Media_Source_Extensions_API) that your browser implements to play video.

**Reply by tenpaiyomi** [reply]
Could you expand on your hardware setup, perhaps?  Maybe that might help isolate some common factor, but I would be surprised if there was something about my particular hardware combination that was causing this for a single web app.  For myself, I've got a 4090 and 7700X.

I'm also aware that Frigate uses the MSE API calls, I went peeking through the DOM and noted the `video` tag feeding in from a `blob`.  As noted, nothing really jumps out at me as to what would cause the issue, but it _only_ happens with Frigate, which just feels a bit too specific to just be mere happenstance or a one off

**Reply by hawkeye217** [reply]
I'm using Firefox 145 on an M1 Mac. My cameras send H.264 video and AAC audio.

**Reply by tenpaiyomi** [reply]
I'm on 146 on Windows.  Everything on mine is coming from Scrypted as H.264 as well.  Checking on my M2 Mini, the GPU barely goes over ~6-7%, so at least in this my case there is definitely _something_ system based.  Any thoughts on what I might be able to check to help potentially provide some more information?

**Reply by hawkeye217** [reply]
Perhaps it's some kind of Windows driver or the way Firefox interacts with it. Do you see the issue when you view a single camera?

**Reply by tenpaiyomi** [reply]
Yes, to a lesser degree.  Viewing a single camera, the GPU video decode is around 60%.  I did notice when viewing one particular camera, the video decode was around 17%, but the actual stats seemed to show the same approximate bitrate and latency.  These are the same model cameras and set up the same, so I'm digging into it see if there is something else to explain why there is a variance between them.
